src/util/sharing_node.h

```diff
@@ -12,340 +12,499 @@ Author: Daniel Poetzl
 #ifndef CPROVER_UTIL_SHARING_NODE_H
 #define CPROVER_UTIL_SHARING_NODE_H
 
-#include <list>
+#define DEBUG
+
+#ifdef DEBUG
+#include <iostream>
+#endif
+
+#include <forward_list>
 #include <map>
 #include <memory>
-#include <cassert>
+#include <type_traits>
 
 #include "invariant.h"
+#include "make_unique.h"
 
-#define _sn_assert(b) INVARIANT(b, "Sharing-node internal invariant")
+#define _sn_assert(b) INVARIANT(b, "Sharing node internal invariant")
 //#define _sn_assert(b)
 
+#define _sn_type_par_decl                                                      \
```

Comment:
Why do all these things start with an underscore? I don't think this matches our internal naming convention.

Reply:
I've now renamed all the macros to be all uppercase and have dropped the leading underscore.